File: packages/coding-agent/src/core/kernel/state-snapshot.ts
// Locations and result shapes for the kernel's persisted user namespace, which
// is revived when a session resumes. The kernel is otherwise spawned fresh on
// resume, leaving the model believing it still has access to variables/imports
// it defined earlier.
//
// Snapshotting is best-effort and per-variable: each top-level name is pickled
// with `dill` independently, so a single unpicklable object (open file, socket,
// GPU tensor, …) is skipped and reported rather than aborting the whole snapshot.
import { lstatSync } from "node:fs";
import { join } from "node:path";

/** Default ceiling on a snapshot payload. Over-cap variables are skipped + reported. */
export const DEFAULT_SNAPSHOT_MAX_BYTES = 256 * 1024 * 1024;
/** Default ceiling for one serialized variable. */
export const DEFAULT_SNAPSHOT_MAX_VARIABLE_BYTES = 16 * 1024 * 1024;

/** Base filename for the kernel snapshot within a session's artifact directory. */
const KERNEL_STATE_BASENAME = "kernel-state";

export type KernelSnapshotFormat = "legacy" | "cas-v2";
export type KernelRestoreSource = "auto" | "current" | "previous" | "legacy";

/** Numeric-only timings and byte counts returned by the Python serializer. */
export interface SnapshotPerformanceMetadata {
	serialization_wall_ms: number | null;
	serialization_cpu_ms: number | null;
	serialized_bytes: number | null;
	write_ms: number | null;
	written_bytes: number | null;
	total_wall_ms: number | null;
}

export interface SnapshotResult {
	/** Top-level names successfully serialized into the payload. */
	saved: string[];
	/** Names that could not be serialized, with a short reason. */
	skipped: { name: string; reason: string }[];
	/** Oversized live variables removed by an explicit compaction snapshot. */
	pruned?: string[];
	/** Legacy-equivalent payload bytes, including outer-container overhead. */
	bytes: number;
	/** Sum of retained independent per-name dill blobs. */
	logicalBytes: number;
	/** Bytes physically written by this snapshot attempt, including CAS metadata. */
	writtenBytes: number;
	format: KernelSnapshotFormat;
	generation?: string;
	/** False for CAS v2 until an explicit legacy export is completed. */
	backwardReadable: boolean;
	metrics?: SnapshotPerformanceMetadata;
	path: string;
}

export interface RestoreResult {
	/** Names successfully revived into the kernel namespace. */
	restored: string[];
	/** Names present in the snapshot that failed to revive, with a short reason. */
	failed: { name: string; reason: string }[];
	format?: KernelSnapshotFormat;
	generation?: string;
	/** Previous-generation restore was explicitly requested. */
	rolledBack?: boolean;
	/** The selected source may omit newer committed or unsaved work. */
	unsavedWorkPossible?: boolean;
	/** Legacy was explicitly selected despite a v2 state root. */
	legacyRecovery?: boolean;
	path: string;
}

export interface SnapshotLegacyExportResult {
	exported: string[];
	bytes: number;
	sourceGeneration: string;
	source: "current" | "previous";
	backwardReadable: true;
	path: string;
}

/** Absolute path to the legacy dill payload within a session's artifact directory. */
export function snapshotPathIn(artifactDir: string): string {
	return join(artifactDir, `${KERNEL_STATE_BASENAME}.dill`);
}

/** Absolute path to the legacy JSON manifest within a session's artifact directory. */
export function manifestPathIn(artifactDir: string): string {
	return join(artifactDir, `${KERNEL_STATE_BASENAME}.json`);
}

/** Session-scoped root for CAS v2 blobs, generations, marker, and pointer. */
export function casSnapshotRootIn(artifactDir: string): string {
	return join(artifactDir, `${KERNEL_STATE_BASENAME}.v2`);
}

/** Derive a sibling CAS root for direct manager users that only supply a legacy path. */
export function casSnapshotRootForLegacyPath(snapshotPath: string): string {
	return snapshotPath.endsWith(".dill") ? `${snapshotPath.slice(0, -".dill".length)}.v2` : `${snapshotPath}.v2`;
}

function pathEntryExists(path: string): boolean {
	try {
		return lstatSync(path, { throwIfNoEntry: false }) !== undefined;
	} catch {
		// Permission and malformed-reparse failures are state, not proof of absence.
		return true;
	}
}

/** True for a legacy payload or any v2 root entry, including broken reparse links. */
export function snapshotStateExistsIn(artifactDir: string): boolean {
	return pathEntryExists(snapshotPathIn(artifactDir)) || pathEntryExists(casSnapshotRootIn(artifactDir));
}

/** Manager-level v2 detection when only explicit paths are available. */
export function casSnapshotStateExists(root: string): boolean {
	return pathEntryExists(root);
}

File: crates/pi-coding-agent/src/core/kernel/state_snapshot.rs
//! Port of packages/coding-agent/src/core/kernel/state-snapshot.ts.
//!
//! Locations and result shapes for the kernel's persisted user namespace, which
//! is revived when a session resumes. The kernel is otherwise spawned fresh on
//! resume, leaving the model believing it still has access to variables/imports
//! it defined earlier.
//!
//! Snapshotting is best-effort and per-variable: each top-level name is pickled
//! with `dill` independently, so a single unpicklable object (open file, socket,
//! GPU tensor, ...) is skipped and reported rather than aborting the whole snapshot.
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

/// Default ceiling on a snapshot payload. Over-cap variables are skipped + reported.
pub const DEFAULT_SNAPSHOT_MAX_BYTES: u64 = 256 * 1024 * 1024;
/// Default ceiling for one serialized variable.
pub const DEFAULT_SNAPSHOT_MAX_VARIABLE_BYTES: u64 = 16 * 1024 * 1024;

/// Base filename for the kernel snapshot within a session's artifact directory.
const KERNEL_STATE_BASENAME: &str = "kernel-state";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum KernelSnapshotFormat {
    Legacy,
    CasV2,
}

impl KernelSnapshotFormat {
    pub fn as_str(self) -> &'static str {
        match self {
            KernelSnapshotFormat::Legacy => "legacy",
            KernelSnapshotFormat::CasV2 => "cas-v2",
        }
    }

    pub fn from_str(value: &str) -> Option<Self> {
        match value {
            "legacy" => Some(KernelSnapshotFormat::Legacy),
            "cas-v2" => Some(KernelSnapshotFormat::CasV2),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum KernelRestoreSource {
    Auto,
    Current,
    Previous,
    Legacy,
}

impl KernelRestoreSource {
    pub fn as_str(self) -> &'static str {
        match self {
            KernelRestoreSource::Auto => "auto",
            KernelRestoreSource::Current => "current",
            KernelRestoreSource::Previous => "previous",
            KernelRestoreSource::Legacy => "legacy",
        }
    }
}

/// Numeric-only timings and byte counts returned by the Python serializer.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct SnapshotPerformanceMetadata {
    pub serialization_wall_ms: Option<f64>,
    pub serialization_cpu_ms: Option<f64>,
    pub serialized_bytes: Option<f64>,
    pub write_ms: Option<f64>,
    pub written_bytes: Option<f64>,
    pub total_wall_ms: Option<f64>,
}

/// One skipped (unserializable) name and a short reason.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SkippedVariable {
    pub name: String,
    pub reason: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SnapshotResult {
    /// Top-level names successfully serialized into the payload.
    pub saved: Vec<String>,
    /// Names that could not be serialized, with a short reason.
    pub skipped: Vec<SkippedVariable>,
    /// Oversized live variables removed by an explicit compaction snapshot.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pruned: Option<Vec<String>>,
    /// Legacy-equivalent payload bytes, including outer-container overhead.
    pub bytes: u64,
    /// Sum of retained independent per-name dill blobs.
    pub logical_bytes: u64,
    /// Bytes physically written by this snapshot attempt, including CAS metadata.
    pub written_bytes: u64,
    pub format: KernelSnapshotFormat,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub generation: Option<String>,
    /// False for CAS v2 until an explicit legacy export is completed.
    pub backward_readable: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metrics: Option<SnapshotPerformanceMetadata>,
    pub path: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreResult {
    /// Names successfully revived into the kernel namespace.
    pub restored: Vec<String>,
    /// Names present in the snapshot that failed to revive, with a short reason.
    pub failed: Vec<SkippedVariable>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub format: Option<KernelSnapshotFormat>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub generation: Option<String>,
    /// Previous-generation restore was explicitly requested.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rolled_back: Option<bool>,
    /// The selected source may omit newer committed or unsaved work.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub unsaved_work_possible: Option<bool>,
    /// Legacy was explicitly selected despite a v2 state root.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub legacy_recovery: Option<bool>,
    pub path: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SnapshotLegacyExportResult {
    pub exported: Vec<String>,
    pub bytes: u64,
    pub source_generation: String,
    pub source: LegacyExportSource,
    pub backward_readable: bool,
    pub path: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum LegacyExportSource {
    Current,
    Previous,
}

impl LegacyExportSource {
    pub fn as_str(self) -> &'static str {
        match self {
            LegacyExportSource::Current => "current",
            LegacyExportSource::Previous => "previous",
        }
    }
}

/// Absolute path to the legacy dill payload within a session's artifact directory.
pub fn snapshot_path_in(artifact_dir: &str) -> String {
    join_path(artifact_dir, &format!("{KERNEL_STATE_BASENAME}.dill"))
}

/// Absolute path to the legacy JSON manifest within a session's artifact directory.
pub fn manifest_path_in(artifact_dir: &str) -> String {
    join_path(artifact_dir, &format!("{KERNEL_STATE_BASENAME}.json"))
}

/// Session-scoped root for CAS v2 blobs, generations, marker, and pointer.
pub fn cas_snapshot_root_in(artifact_dir: &str) -> String {
    join_path(artifact_dir, &format!("{KERNEL_STATE_BASENAME}.v2"))
}

/// Derive a sibling CAS root for direct manager users that only supply a legacy path.
pub fn cas_snapshot_root_for_legacy_path(snapshot_path: &str) -> String {
    match snapshot_path.strip_suffix(".dill") {
        Some(prefix) => format!("{prefix}.v2"),
        None => format!("{snapshot_path}.v2"),
    }
}

fn join_path(dir: &str, name: &str) -> String {
    PathBuf::from(dir).join(name).to_string_lossy().into_owned()
}

fn path_entry_exists(path: &str) -> bool {
    match std::fs::symlink_metadata(Path::new(path)) {
        Ok(_) => true,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => false,
        // Permission and malformed-reparse failures are state, not proof of absence.
        Err(_) => true,
    }
}

/// True for a legacy payload or any v2 root entry, including broken reparse links.
pub fn snapshot_state_exists_in(artifact_dir: &str) -> bool {
    path_entry_exists(&snapshot_path_in(artifact_dir)) || path_entry_exists(&cas_snapshot_root_in(artifact_dir))
}

/// Manager-level v2 detection when only explicit paths are available.
pub fn cas_snapshot_state_exists(root: &str) -> bool {
    path_entry_exists(root)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn snapshot_paths_are_derived_from_the_artifact_dir() {
        let dir = if cfg!(windows) { "C:\\artifacts" } else { "/artifacts" };
        assert!(snapshot_path_in(dir).ends_with("kernel-state.dill"));
        assert!(manifest_path_in(dir).ends_with("kernel-state.json"));
        assert!(cas_snapshot_root_in(dir).ends_with("kernel-state.v2"));
    }

    #[test]
    fn cas_root_for_legacy_path_strips_the_dill_suffix() {
        assert_eq!(cas_snapshot_root_for_legacy_path("/a/kernel-state.dill"), "/a/kernel-state.v2");
        assert_eq!(cas_snapshot_root_for_legacy_path("/a/kernel-state"), "/a/kernel-state.v2");
    }

    #[test]
    fn missing_entries_are_absent_and_present_entries_exist() {
        let dir = std::env::temp_dir().join(format!("pi-snapshot-test-{}", std::process::id()));
        std::fs::create_dir_all(&dir).unwrap();
        let dir_string = dir.to_string_lossy().into_owned();
        assert!(!snapshot_state_exists_in(&dir_string));
        std::fs::write(snapshot_path_in(&dir_string), b"x").unwrap();
        assert!(snapshot_state_exists_in(&dir_string));
        assert!(cas_snapshot_state_exists(&snapshot_path_in(&dir_string)));
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn snapshot_format_round_trips_wire_names() {
        assert_eq!(KernelSnapshotFormat::CasV2.as_str(), "cas-v2");
        assert_eq!(KernelSnapshotFormat::from_str("legacy"), Some(KernelSnapshotFormat::Legacy));
        assert_eq!(KernelSnapshotFormat::from_str("nope"), None);
        assert_eq!(KernelRestoreSource::Previous.as_str(), "previous");
    }
}
